### DISCUSSION TITLE
Bootstrap on Mac OS

### DISCUSSION THREAD

**Discussion by ashkan-leo** [malob/nix-config#6] [Q&A] [answered]
Hello!

This repository is very inspiring. Thanks for sharing it with us!
Would you mind adding some notes on how to bootstrap a macOS system? 
I am trying to enable flake on my macOS without any luck so far.

**Comment by malob** [comment, marked as answer]
It's on my list at some point to add some notes to the readme about bootstrapping. I haven't added that yet because I personally have never bootstrapped my current config from scratch on a macOS system (I switch to the current flake setup on an existing system), and I don't want to add "official" instructions to the readme until I've actually tested them to ensure they work. The [GitHub workflow](https://github.com/malob/nixpkgs/blob/master/.github/workflows/ci.yml) on this repo is a pretty good place to look to get the basic picture of what the steps required look like though.

Here are some additional notes that should be helpful:

* You'll of course want Nix (multi-user) installed on your system. The instructions for doing that are well covered by the Nix manual.
* With Nix installed, you can upgrade to a version of Nix that supports flakes by running `nix-env -iA nixpkgs.nixFlakes`.
* To actually used Flake related commands, you need to add `experimental-features = nix-command flakes` to your `nix.conf` file, or run Nix commands with a flag to enable the Flake features, `nix --experimental-features 'nix-command flakes' [...]`.
* `nix-darwin` requires that you have a `/run` directory. If that's not already setup on your system you can run the following commands to create it:
  ```bash
  echo -e "run\tprivate/var/run" | sudo tee -a /etc/synthetic.conf >/dev/null
  /System/Library/Filesystems/apfs.fs/Contents/Resources/apfs.util -B 2>/dev/null
  ```
* With all that done you can then build a `nix-darwin` configuration with `nix build .#darwinConfigurations.[name].system`, then activate that configuration by running `./result/sw/bin/darwin-rebuild switch --flake .#[name]` (both of which should be run from the root of the repository with your configuration).
  * When running `darwin-rebuild switch` for the first time, you might get some non-fatal errors saying that files under `/etc` can't be linked, e.g., `/etc/shells`, since they already exist. You'll want to delete those and run `darwin-rebuild switch` again.
  * You'll notice that in the GitHub workflow I first build and switch to the `bootstrap` configuration I made. I do that first because it adds some basic Nix configuration to my environment first, like additional binary caches, which make building my full config for the first time faster. It's not essential or anything though.
* After doing all that, and restarting your shell, your `nix-darwin` environment/configuration should all be setup and working, and you can update/change your config and run `darwin-rebuild switch --flake .#[name]`, to apply those changes.

**Reply by malob** [reply]
Hope this helps, feel free to ask followup questions :)

**Reply by lucamaraschi** [reply]
@malob fantastic job!I was wondering if `homebrew` should be install by nix-darwin or if I need to install it separately. I saw it in the `flake` but it doesn't seems to get installed automatically...

**Reply by malob** [reply]
Thanks!

The current `homebrew` module in `nix-darwin` does not handle Homedrew installation, so you'll need to install it yourself.

**Reply by lucamaraschi** [reply]
Thanks a lot @malob!

**Reply by r17x** [reply]
@malob  I have been follow your guide, after do `/result/sw/bin/darwin-rebuild switch --flake .#[name]` and restart SHELL. 

`darwin-rebuild` not found in my shell 

Do you have any idea for that ?

**Reply by malob** [reply]
@ri7nz, hmmm, hard to know for sure. Here are a few ideas off the top of my head:
* Just to sanity check, does the issue persist after reboot?
* If so:
  * Is `nix` available in your environment? If it isn't then the issue seems to be that your shell isn't sourcing the Nix environment at all.
  * Are `nix` and `darwin-rebuild` available in your shell if you open a new terminal and run `exec /run/current-system/sw/bin/fish`? If so, then it might be the case then you might need to run `chsh` to set the default shell that terminal emulators on your system will use. Try running `chsh -s /run/current-system/sw/bin/fish` to set Fish as you default shell. You'll want to select one of the shells installed by your Nix config which will be listed in `/etc/shells` under `# List of shells managed by nix`.

**Reply by r17x** [reply]
@malob  Thank you. 

This is my environment state
* `nix` available in my shell when first run my machine
* `darwin-rebuild` not found
So, I follow your comment:
* just run `exec /run/current-system/sw/bin/fish` and `chsh -s /run/current-system/sw/bin/fish`

It's work correctly.

But, I have some question again @malob .
```console
❯ echo $NIX_PATH
darwin-config=/Users/r17/.nixpkgs/darwin-configuration.nix nixpkgs=/Users/r17/.config/nixpkgs/nixpkgs.nix /nix/var/nix/profiles/per-user/root/channels /Users/r17/.nix-defexpr/channels
```
Why `darwin-config` reference to some path, even though the file doesn't exist.

**Reply by malob** [reply]
> But, I have some question again @malob .
>
> ❯ echo $NIX_PATH
darwin-config=/Users/r17/.nixpkgs/darwin-configuration.nix nixpkgs=/Users/r17/.config/nixpkgs/nixpkgs.nix /nix/var/nix/profiles/per-user/root/channels /Users/r17/.nix-defexpr/channels
Why darwin-config reference to some path, even though the file doesn't exist.
>
> Why darwin-config reference to some path, even though the file doesn't exist.

@ri7nz, my system also has `darwin-config` listed on my `NIX_PATH` pointing to the same file. I believe this is the default location where `nix-darwin` looks for it's configuration when not using Flakes. 

In a (non-Flakes) traditional setup `darwin-rebuild` builds the configuration pointed to by `darwin-config` in the `NIX_PATH`, and you can use the `nix.nixPath` option to explicitly set is to point to a different location. But for Flakes users this doesn't matter. Perhaps `nix-darwin` should be updated so that `darwin-config` isn't added to `nix.nixPath` when using Flakes just to keep things a little tidier ¯\_(ツ)_/¯

**Reply by r17x** [reply]
@malob  Nevermind for my question before. it's work correctly right now. Thank you 


Anyway, do you know how to make linux configuration in flake.nix since flake.nix have a darwinConfiguration ? I want to run current nixpkgs config to WSL2

**Reply by malob** [reply]
I've only played around with WSL2 a little bit, so I'm not that knowledgeable here.

One thing I'm almost certain you can do is install Nix in a non-NixOS distribution, e.g., Ubuntu, and use a standalone `home-manager` configuration to manage things. You can create a `homeConfigurations` output in your flake, like I have here: https://github.com/malob/nixpkgs/blob/1540c7b83ae86da9c549593a21744a60e2e41252/flake.nix#L137-L153

For more details have a look at their documentation for standalone configurations using flakes: https://nix-community.github.io/home-manager/index.html#sec-flakes-standalone

It might also be possible to setup NixOS with WSL2. If that's the case you could create a `nixosConfigurations` output in your flake: https://nixos.wiki/wiki/Flakes#Using_nix_flakes_with_NixOS

Though I wouldn't expect there to be much benefit to using NixOS with WSL2 vs Ubuntu with Nix installed and a standalone `home-manager` env.

**Reply by malob** [reply]
Looks like a proper NixOS install is possible with WSL2. See here: https://github.com/nix-community/NixOS-WSL
